### PR TITLE
Instructions for `o3.Linear`; zero multiplicities fix

### PR DESCRIPTION
**Pull request by Linux-cpp-lisp**
## Description

- Allow instructions to be specified to `o3.Linear` to mix only certain irreps
- Weight view methods for `o3.Linear`
- Fix zero multiplicities for TensorProduct
- Make weight views support `shared_weights=False`

## Motivation and Context
Previously, there was no public way to interpret the `.weight` of an `o3.Linear`, since the order of paths was not exposed.

## How Has This Been Tested?
e3nn tests

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds or improves functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation improvement (updates to user guides, docstrings, or developer docs)

## Checklist:

- [X] I have read the [**CONTRIBUTING**](https://github.com/e3nn/e3nn/blob/main/CONTRIBUTING.md) document.
- [X] My code follows the code style of this project.
- [X] I have updated the documentation (if relevant).
- [X] I have added tests that cover my changes (if relevant).
- [X] All new and existing tests passed.
- [X] I have updated the [Changelog](https://github.com/e3nn/e3nn/blob/main/ChangeLog.md).

### CODE REVIEW

File: e3nn/o3/_linear.py [hunk 1]
```diff
@@ -185,45 +297,48 @@ def _codegen_linear(
...
+            (() if shared_weights else (-1,)) + ins.path_shape
         )
         flat_weight_index += path_nweight
 
-        ein_out = torch.einsum(f"{z}uw,zui->zwi", w, x_list[i_in])
-        alpha = 1.0 / math.sqrt(mul_ir_in.mul * sum(1 if i_out_this == i_out else 0 for _, i_out_this in instr))
+        ein_out = torch.einsum(f"{z}uw,zui->zwi", w, x_list[ins.i_in])
+        alpha = 1.0 / math.sqrt(
+            mul_ir_in.mul * sum(
+                1 if other_ins.i_out == ins.i_out else 0
```
Comment (mariogeiger) [inline e3nn/o3/_linear.py:330]: Maybe I could make a remark in the doc that `Linear(10x0, 0)` is not equivalent to `Linear(2x0 + 8x0, 0)`

I seriously wonder if we should remove the `simplify` operations

**Review by mariogeiger — APPROVED** [review]
Very nice code. I love the attention to detail 👍

File: e3nn/o3/_linear.py [hunk 2]
```diff
@@ -70,17 +89,54 @@ def __init__(
...
         self._optimize_einsums = _optimize_einsums if _optimize_einsums is not None else opt_defaults['optimize_einsums']
         del opt_defaults
 
+        # == Instructions ==
+        if instructions is None:
+            # In the fully connected case, everything goes everywhere anyway,
+            # so we can simplify the irreps and maybe save some compute:
+            self.irreps_in = self.irreps_in.simplify()
+            self.irreps_out = self.irreps_out.simplify()
+            # ^ we can't do this in the general case; see the block sparse example.
```
Comment (mariogeiger) [inline e3nn/o3/_linear.py:107]:
```suggestion
```

File: e3nn/o3/_linear.py [hunk 3]
```diff
@@ -41,6 +51,14 @@ class Linear(CodeGenMixin, torch.nn.Module):
...
     >>> lin.weight_numel
     160
+
+    Create a "block sparse" linear that does not combine two different groups of scalars;
+    note that the number of weights is 4*4 + 3*3 = 25:
+
+    >>> lin = Linear("4x0e + 3x0e", "4x0e + 3x0e", instructions=[(0, 0), (1, 1)])
+    >>> lin.weight_numel
+    25
+
```
Comment (mariogeiger) [inline e3nn/o3/_linear.py:61]:
```suggestion

    Be careful, the two following operations are not normalized the same way
    
    >>> lin1 = Linear("10x0e", "0e")
    >>> lin2 = Linear("3x0e + 7x0e", "0e")
    >>> with torch.no_grad():
    ...     lin1.weight.fill_(1.0)
    ...     lin2.weight.fill_(1.0)
    >>> x = torch.arange(10.0)
    >>> (lin1(x) - lin2(x)).abs().item > 0.1
    True
```

Comment (mariogeiger) [reply e3nn/o3/_linear.py:61]: (code not tested)

File: e3nn/o3/_linear.py [hunk 4]
```diff
@@ -98,12 +154,11 @@ def __init__(
...
         # == Compute output mask ==
         if self.irreps_out.dim > 0:
-            # an output is connected to an input if one exists with the same L, p
             output_mask = torch.cat([
-                torch.ones(mul * ir.dim)
-                if any(ir_in == ir for _, ir_in in self.irreps_in)
-                else torch.zeros(mul * ir.dim)
-                for _, (mul, ir) in enumerate(self.irreps_out)
+                torch.ones(mul_ir.dim)
+                if any(ins.i_out == i_out for ins in self.instructions)
```
[the remark this answers is not in the document]
Comment (Linux-cpp-lisp) [reply e3nn/o3/_linear.py:159]: not sure I understand what you want to change

File: e3nn/o3/_linear.py [hunk 5]
```diff
@@ -185,45 +297,48 @@ def _codegen_linear(
 
     out_list = []
 
-    instr = [
-        (i_in, i_out)
-        for i_in, (_, ir_in) in enumerate(irreps_in)
-        for i_out, (_, ir_out) in enumerate(irreps_out)
-        if ir_in == ir_out
-    ]
+    # TODO: ?? "fuse" instructions to make more efficient? I.e.:
```
[the remark this answers is not in the document]
Comment (Linux-cpp-lisp) [reply e3nn/o3/_linear.py:300]: yes agree, I'll remove the note

File: e3nn/o3/_linear.py (hunk 1 above)
Comment (Linux-cpp-lisp) [reply e3nn/o3/_linear.py:330]: aha yes, this didn't occur to me — it should be removed and there should definitely be a note in the docs.

Maybe also more discussion somewhere in the docs in general that it can be very good to call `simplify()` yourself to get more efficient networks... maybe it is time to start an "FAQ" or a "Tips" page in the docs?

File: e3nn/o3/_linear.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (mariogeiger) [reply e3nn/o3/_linear.py:181]: The mask should be one for the outputs that have a strictly positive number of inputs

File: e3nn/o3/_linear.py (hunk 1 above)
Comment (mariogeiger) [reply e3nn/o3/_linear.py:353]: Yes tips and/or faq is a good idea

**Review by mariogeiger — APPROVED** [review]
